server/src/main/java/org/opensearch/common/cache/stats/StatsHolder.java

```diff
@@ -0,0 +1,214 @@
+    // which have been entirely deleted, for example in an async removal listener.
+    public void decrementSizeInBytes(List<String> dimensionValues, long amountBytes) {
+        internalIncrement(dimensionValues, (node) -> node.decrementSizeInBytes(amountBytes), false);
+    }
+
+    public void incrementEntries(List<String> dimensionValues) {
+        internalIncrement(dimensionValues, DimensionNode::incrementEntries, true);
+    }
+
+    public void decrementEntries(List<String> dimensionValues) {
+        internalIncrement(dimensionValues, DimensionNode::decrementEntries, false);
+    }
+
+    /**
+     * Reset number of entries and memory size when all keys leave the cache, but don't reset hit/miss/eviction numbers.
+     * This is in line with the behavior of the existing API when caches are cleared.
+     */
+    public void reset() {
+        resetHelper(statsRoot);
+    }
+
+    private void resetHelper(DimensionNode current) {
+        current.resetSizeAndEntries();
+        for (DimensionNode child : current.children.values()) {
+            resetHelper(child);
+        }
+    }
+
+    public long count() {
+        // Include this here so caches don't have to create an entire CacheStats object to run count().
+        return statsRoot.getEntries();
+    }
+
+    private void internalIncrement(List<String> dimensionValues, Consumer<DimensionNode> adder, boolean createNodesIfAbsent) {
+        assert dimensionValues.size() == dimensionNames.size();
+        internalIncrementHelper(dimensionValues, statsRoot, 0, adder, createNodesIfAbsent);
+    }
+
+    /**
+     * Use the incrementer function to increment/decrement a value in the stats for a set of dimensions.
+     * If createNodesIfAbsent is true, and there is no stats for this set of dimensions, create one.
+     * Returns true if the increment was applied, false if not.
+     */
+    private boolean internalIncrementHelper(
+        List<String> dimensionValues,
+        DimensionNode node,
+        int depth, // Pass in the depth to avoid having to slice the list for each node.
+        Consumer<DimensionNode> adder,
+        boolean createNodesIfAbsent
+    ) {
+        if (depth == dimensionValues.size()) {
+            // This is the leaf node we are trying to reach
+            adder.accept(node);
+            return true;
+        }
+
+        DimensionNode child = node.getChild(dimensionValues.get(depth));
+        if (child == null) {
+            if (createNodesIfAbsent) {
+                // If we have to create a new node, obtain the lock first
+                boolean createMapInChild = depth < dimensionValues.size() - 1;
+                lock.lock();
+                try {
+                    child = node.createChild(dimensionValues.get(depth), createMapInChild);
+                } finally {
+                    lock.unlock();
+                }
```

Comment:
Okay I initially thought this lock is not helping at all, since the createChild function itself is synchronized. But, this helps add synchronization across potentially different concurrent maps during creation and removal